File: 5th Semester/INF318/UNCICLO.py
# Mostrar Factoriales

def mostrarFactoriales(n):
	f = 1
	i = 1
	while(i<=n):
		f = f*i
		print(str(i)+"!="+str(f))
		i = i + 1
		
def mostrarFactoriales_rec(n):
	if(n==0 or n==1):
		print(str(n)+"!=1")
		return 1
	f = n * mostrarFactoriales_rec(n-1)
	print(str(n)+"!="+str(f))
	return f
	
# Primo

def primo(n):
	if(n==1):
		return False
	i = 2
	while(i<=n/2):
		if(n%i==0):
			return False
		i = i + 1
	return True
def primo_rec(n):
	return primo_recp(n,2)

def primo_recp(n,i):
	if(n==1):
		return False
	if(i>n/2):
		return True
	if(n%i==0):
		return False
	return primo_recp(n,i+1)

# Mostrar Primos	

def mostrarPrimos(a,b):
	while(a<=b):
		if(primo(a)):
			print(a)
		a = a + 1
		
def mostrarPrimos_rec(a,b):
	if(a<=b):
		if(primo(a)):
			print(a)
		mostrarPrimos_rec(a+1,b)
		
# Mostrar Divisores Descendientes
		
def mostrarDivisoresDesc(n):
	i = n
	while(i>=1):
		if(n%i==0):
			print(i)
		i = i - 1
		
		
def mostrarDivisoresDesc_rec(n):
	mostrarDivisoresDesc_recp(n,1)
	
def mostrarDivisoresDesc_recp(n,i):
	if(i<=n):
		mostrarDivisoresDesc_recp(n,i+1)
		if(n%i==0):
			print(i)
			
# Coeficiente Binomial
			
def fact(n):
	if(n==0 or n==1):
		return 1
	return n*fact(n-1)
			
def comb(n, r):
	return fact(n)/(fact(n-r)*fact(r))
	
def mostrarCoefBin(n):
	i = 0
	while (i <= n):
		print(comb(n,i))
		i += 1
		
def mostrarCoefBin_rec(n):
	mostrarCoefBinR(n,0)
	
def mostrarCoefBinR(n, i):
	if (i <= n):
		print(comb(n,i))
		mostrarCoefBinR(n, i+1)

File: 5th Semester/INF318/UNCICLO.pl
% Mostrar  Factoriales

mostrarFactorial:- read(N),mostrarFactorial(N,_).

mostrarFactorial(0,1):-!.
mostrarFactorial(N,R):- N1 is N-1,
						mostrarFactorial(N1,R1),
						R is N*R1,
						write(N),
						write("!="),
						write(R),nl.

% Primo
						
primo:- read(N),primo(N).

primo(N):- primo(N,2).
primo(N,I):- I > N // 2.
primo(N,I):- N mod I =\= 0,
			I1 is I+1,
			primo(N,I1).
			
% Mostrar Divisores Descendientes

mostrarDivisoresDesc:- read(N),divisores(N,1).

divisores(N,I):- I>N,!.
divisores(N,I):- N mod I =:= 0,
				I1 is I+1,
				divisores(N,I1),
				write(I),nl.
divisores(N,I):- I1 is I+1,divisores(N,I1).

% Mostrar Primos Intervalo

mostrarPrimo(A,B):- A>B, !.
mostrarPrimo(A,B):- primo(A),
					write(A),write(" "),
					A1 is A+1,
					mostrarPrimo(A1,B).
mostrarPrimo(A,B):- A1 is A+1, mostrarPrimo(A1,B).

% Mostrar Coeficientes Binomiales
fact(N) :- fact(N,N).
fact(N,F):- N<2, F is 1, !.
fact(N,F):- N1 is N-1,fact(N1,F1),F is N*F1.


comb(N,R,C):- X is N-R,
			fact(N,C1),
			fact(X,C2),
			fact(R,C3),
			C is (C1/(C2*C3)).
			
mostrarCoefBin(N):- mostrarCoefBin(N,0).
mostrarCoefBin(N,I):- I>N, !.
mostrarCoefBin(N,I):- comb(N,I,C),
					write(C),write(" "),
					I1 is I+1,
					mostrarCoefBin(N,I1).
